### PR TITLE
Fix integrator parameter indexing, enable tests

### PR DESCRIPTION
**Pull request by AayushSabharwal**
Contains the fixes of #272 without the change that broke tests

### CODE REVIEW

File: test/downstream/integrator_indexing.jl [hunk 1]
```diff
@@ -54,8 +54,8 @@ step!(integrator, 100.0, true)
 @test integrator[lorenz1.x] isa Real
 @test integrator[t] isa Real
 @test integrator[α] isa Real
-# @test integrator[γ] isa Real
-# @test integrator[γ] == 2.0
+@test integrator[γ] isa Real
+@test integrator[γ] == 2.0
```
Comment (ChrisRackauckas) [inline test/downstream/integrator_indexing.jl:59]: Test the values and the cases that @YingboMa found broke.

Comment (ChrisRackauckas) [reply test/downstream/integrator_indexing.jl]: no, it doesn't cover the array of symbols handling.

Comment (AayushSabharwal) [reply test/downstream/integrator_indexing.jl]: So something like `sol[[lorenz1.x, lorenz2.x]]` is required?

Comment (ChrisRackauckas) [reply test/downstream/integrator_indexing.jl]: yes

**Comment by ChrisRackauckas** [comment]
Actually now that I'm looking at it, https://github.com/SciML/SciMLBase.jl/pull/279#issuecomment-1277719422

@YingboMa it looks like the old behavior was broken and the new behavior is right? `sol[x]` is supposed to give the time series of `x`. `x` is a length 3 vector, so it should give a time series of length 3 vectors. In a sense, it should be the same result as `sol[1:3,:]`. So the changed version looks correct and it looks like it was broken before and just not caught/tested. 

This is important for the PDE case, where @xtalax would expect `sol[u(t,x,y)]` to give a time series of the matrices for `u`. 

@AayushSabharwal you should add a test case here with a symbolic array like in @YingboMa 's example.

Instead of returning `Vector{Vector}`, the returns on this kind of thing should be a `RecursiveArrayTools.DiffEqArray` so that the time points are retained. See how the vector indexing is handled in the non-symbolic case.

**Comment by AayushSabharwal** [comment]
Right now:
```julia
julia> @variables t x(t)[1:2] = [1., 2.]
julia> D = Differential(t)
julia> eqs = [D(x[1]) ~ x[2], x[2] ~ 3t]
julia> sys = ODESystem(eqs, t, [x...], []; name = :sys) |> structural_simplify
julia> prob = ODEProblem(sys, [], (0.0, 5.0))
julia> sol = solve(prob, Tsit5())
julia> sol[x]
2-element Vector{Vector{Float64}}:
 [1.0, 1.000000015, 1.0000018149999998, 1.000184815, 1.0185148149999999, 2.8518148149999987, 38.499999999999964]
 [0.0, 0.0003, 0.0032999999999999995, 0.03329999999999999, 0.3332999999999999, 3.3332999999999986, 15.0]
julia> sol[1:2]
retcode: Success
Interpolation: specialized 4th order "free" interpolation
t: 2-element Vector{Float64}:
 0.0
 9.999999999999999e-5
u: 2-element Vector{Vector{Float64}}:
 [1.0]
 [1.000000015]
julia> sol[1:2, :]
ERROR: BoundsError: attempt to access 1×7 RecursiveArrayTools.VectorOfArray{Float64, 2, Vector{Vector{Float64}}} at index [1:2, 1:7]
Stacktrace:
 [1] throw_boundserror(A::RecursiveArrayTools.VectorOfArray{Float64, 2, Vector{Vector{Float64}}}, I::Tuple{UnitRange{Int64}, Base.Slice{Base.OneTo{Int64}}})
   @ Base ./abstractarray.jl:703
 [2] checkbounds
   @ ./abstractarray.jl:668 [inlined]
 [3] _getindex
   @ ./multidimensional.jl:874 [inlined]
 [4] getindex
   @ ./abstractarray.jl:1241 [inlined]
 [5] getindex(::ODESolution{Float64, 2, Vector{Vector{Float64}}, Nothing, Nothing, Vector{Float64}, Vector{Vector{Vector{Float64}}}, ODEProblem{Vector{Float64}, Tuple{Float64, Float64}, true, SciMLBase.NullParameters, ODEFunction{true, SciMLBase.AutoSpecialize, FunctionWrappersWrappers.FunctionWrappersWrapper{Tuple{FunctionWrappers.FunctionWrapper{Nothing, Tuple{Vector{Float64}, Vector{Float64}, SciMLBase.NullParameters, Float64}}, FunctionWrappers.FunctionWrapper{Nothing, Tuple{Vector{ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}, Vector{ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}, SciMLBase.NullParameters, Float64}}, FunctionWrappers.FunctionWrapper{Nothing, Tuple{Vector{ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}, Vector{Float64}, SciMLBase.NullParameters, ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}}, FunctionWrappers.FunctionWrapper{Nothing, Tuple{Vector{ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}, Vector{ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}, SciMLBase.NullParameters, ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}}}, false}, LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Vector{Symbol}, Symbol, Vector{Any}, ModelingToolkit.var"#487#generated_observed#469"{Bool, ODESystem, Dict{Any, Any}}, Nothing, ODESystem}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, SciMLBase.StandardODEProblem}, Tsit5{typeof(OrdinaryDiffEq.trivial_limiter!), typeof(OrdinaryDiffEq.trivial_limiter!), Static.False}, OrdinaryDiffEq.InterpolationData{ODEFunction{true, SciMLBase.AutoSpecialize, FunctionWrappersWrappers.FunctionWrappersWrapper{Tuple{FunctionWrappers.FunctionWrapper{Nothing, Tuple{Vector{Float64}, Vector{Float64}, SciMLBase.NullParameters, Float64}}, FunctionWrappers.FunctionWrapper{Nothing, Tuple{Vector{ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}, Vector{ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}, SciMLBase.NullParameters, Float64}}, FunctionWrappers.FunctionWrapper{Nothing, Tuple{Vector{ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}, Vector{Float64}, SciMLBase.NullParameters, ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}}, FunctionWrappers.FunctionWrapper{Nothing, Tuple{Vector{ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}, Vector{ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}, SciMLBase.NullParameters, ForwardDiff.Dual{ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Float64}, Float64, 1}}}}, false}, LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Vector{Symbol}, Symbol, Vector{Any}, ModelingToolkit.var"#487#generated_observed#469"{Bool, ODESystem, Dict{Any, Any}}, Nothing, ODESystem}, Vector{Vector{Float64}}, Vector{Float64}, Vector{Vector{Vector{Float64}}}, OrdinaryDiffEq.Tsit5Cache{Vector{Float64}, Vector{Float64}, Vector{Float64}, OrdinaryDiffEq.Tsit5ConstantCache{Float64, Float64}, typeof(OrdinaryDiffEq.trivial_limiter!), typeof(OrdinaryDiffEq.trivial_limiter!), Static.False}}, DiffEqBase.DEStats}, ::UnitRange{Int64}, ::Colon)
   @ SciMLBase ~/d/Julia/SciML/SciMLBase.jl/src/solutions/solution_interface.jl:41
 [6] top-level scope
   @ REPL[23]:1
```

**Comment by ChrisRackauckas** [comment]
That's a pretty big regression then if `1:2` indexing isn't working.

**Comment by ChrisRackauckas** [comment]
```julia
using OrdinaryDiffEq
function lorenz!(du,u,p,t)
 du[1] = 10.0(u[2]-u[1])
 du[2] = u[1]*(28.0-u[3]) - u[2]
 du[3] = u[1]*u[2] - (8/3)*u[3]
end
u0 = [1.0;0.0;0.0]
tspan = (0.0,100.0)
prob = ODEProblem(lorenz!,u0,tspan)
sol = solve(prob,Tsit5())

sol[1:2,:]
sol[1:2,1:5]
sol[1:2,[1,3,5]]
sol[1,[1,3,5]]
sol[[1,2],[1,3,5]]
```

on master, those all work. Add those indexing results to the tests first.

**Comment by xtalax** [comment]
Yes, thank you this looks like a good test

**Comment by AayushSabharwal** [comment]
> ```julia
> using OrdinaryDiffEq
> function lorenz!(du,u,p,t)
>  du[1] = 10.0(u[2]-u[1])
>  du[2] = u[1]*(28.0-u[3]) - u[2]
>  du[3] = u[1]*u[2] - (8/3)*u[3]
> end
> u0 = [1.0;0.0;0.0]
> tspan = (0.0,100.0)
> prob = ODEProblem(lorenz!,u0,tspan)
> sol = solve(prob,Tsit5())
> 
> sol[1:2,:]
> sol[1:2,1:5]
> sol[1:2,[1,3,5]]
> sol[1,[1,3,5]]
> sol[[1,2],[1,3,5]]
> ```
> on master, those all work. Add those indexing results to the tests first.

All of these work. The symbolic stuff seems to be the issue.

**Comment by AayushSabharwal** [comment]
I think there is an issue with the MWE I posted above, the following works too:
```julia
julia> @variables t u(t)[1:3]=[1.0,0.0,0.0]
julia> D = Differential(t)
julia> eqs = [
           D(u[1]) ~ 10.0(u[2] - u[1])
           D(u[2]) ~ u[1]*(28.0-u[3]) - u[2]
           D(u[3]) ~ u[1]*u[2] - 8/3*u[3]
       ]
julia> sys = ODESystem(eqs, t, [u...], []; name=:sys)
julia> prob = ODEProblem(sys, [], (0.0,100.0))
julia> sol = solve(prob, Tsit5())

sol[1:2,:]
sol[1:2,1:5]
sol[1:2,[1,3,5]]
sol[1,[1,3,5]]
sol[[1,2],[1,3,5]]
```

**Comment by ChrisRackauckas** [comment]
yes, but the point is to add this stuff to the tests, not just post in the PR that it probably works today.

File: test/downstream/integrator_indexing.jl [hunk 2]
```diff
@@ -54,8 +54,24 @@ step!(integrator, 100.0, true)
...
+@test integrator[[lorenz1.x, lorenz2.x]] isa Vector{Float64}
```
Comment (ChrisRackauckas) [inline test/downstream/integrator_indexing.jl:59]: also test that the size is correctly n, and length of first is 2.

File: test/downstream/symbol_indexing.jl [hunk 3]
```diff
@@ -66,8 +66,26 @@ sol = solve(prob, Rodas4())
 @test sol[α] isa Vector
 @test sol[α, 3] isa Float64
 @test length(sol[α, 5:10]) == 6
-# @test sol[γ] isa Real
-# @test sol[γ] == 2.0
+@test sol[γ] isa Real
+@test sol[γ] == 2.0
+@test sol[[lorenz1.x, lorenz2.x]] isa Vector{Vector{Float64}}
+@test length(sol[[lorenz1.x, lorenz2.x]]) == 2
```
Comment (ChrisRackauckas) [inline test/downstream/symbol_indexing.jl:72]: this is incorrect, it should be length of the time series.

File: test/downstream/symbol_indexing.jl [hunk 4]
```diff
@@ -66,8 +66,26 @@ sol = solve(prob, Rodas4())
...
+@test all(length.(sol[[lorenz1.x, lorenz2.x]]) .== size(sol, 2))
```
Comment (ChrisRackauckas) [inline test/downstream/symbol_indexing.jl:73]: this is incorrect, each of them should be length 2

File: test/downstream/symbol_indexing.jl [hunk 5]
```diff
@@ -66,8 +66,26 @@ sol = solve(prob, Rodas4())
...
+@test sol[[γ, lorenz1.σ]] isa Vector{Float64}
+@test length(sol[[γ, lorenz1.σ]]) == 2
```
Comment (ChrisRackauckas) [inline test/downstream/symbol_indexing.jl:75]: this is incorrect.

File: test/downstream/symbol_indexing.jl [hunk 6]
```diff
@@ -66,8 +66,26 @@ sol = solve(prob, Rodas4())
...
+@variables q(t)[1:2] = [1.0, 2.0]
+eqs = [D(q[1]) ~ 2q[1]
+       D(q[2]) ~ 2.0]
+@named sys2 = ODESystem(eqs, t, [q...], [])
+sys2_simplified = structural_simplify(sys2)
+prob2 = ODEProblem(sys2, [], (0.0, 5.0))
+sol2 = solve(prob2, Tsit5())
+
+@test sol2[q] isa Vector{Vector{Float64}}
+@test length(sol2[q]) == length(q)
```
Comment (ChrisRackauckas) [inline test/downstream/symbol_indexing.jl:86]: yeah this all seems to be "correctly" testing that the behavior is incorrect.

**Comment by ChrisRackauckas** [comment]
Supersceded by https://github.com/SciML/SciMLBase.jl/pull/301

**Pull request closed without merging**